### PR TITLE
Improve perf and memory usage with reuse cache by slicing inputs till token idx for 1st token generation.

### PR DESCRIPTION
**Pull request by puneeshkhanna**
Reuse cache pre allocates the KV cache. This can be further leveraged to slice the inputs till the token idx for the 1st token inference and improve perf and memory usage by running till the required seq len only rather than running full padded static input (input + max new tokens length). Memory usage and perf will vary according to the actual input seq len but will be better than running full static shapes.

Use torch.zeros to initialize KV cache rather than uninitialized torch.empty.

Also fix bool dtypes default values.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
Llama 7B perf results without and with fix as below:
python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-7b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 1 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
----------------------------------------------------------------------
Throughput (including tokenization) = 131.44301810809804 tokens/second
Memory allocated                    = 12.62 GB
Max memory allocated                = 12.63 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.5840961140056606 seconds
----------------------------------------------------------------------
With fix
----------------------------------------------------------------------
Throughput (including tokenization) = 130.75967053114542 tokens/second
Memory allocated                    = 12.61 GB
Max memory allocated                = 12.62 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 2.1798126350040548 seconds
----------------------------------------------------------------------

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-7b-hf/ --max_new_tokens 1000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 1 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
---------------------------------------------------------------------
Throughput (including tokenization) = 138.1225755548041 tokens/second
Memory allocated                    = 13.19 GB
Max memory allocated                = 13.23 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 10.586436815996422 seconds
---------------------------------------------------------------------
With fix
----------------------------------------------------------------------
Throughput (including tokenization) = 139.05396023775887 tokens/second
Memory allocated                    = 13.08 GB
Max memory allocated                = 13.1 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 8.58467547201144 seconds
----------------------------------------------------------------------

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-7b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 128 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
---------------------------------------------------------------------
Throughput (including tokenization) = 7084.259833894209 tokens/second
Memory allocated                    = 20.68 GB
Max memory allocated                = 22.21 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 17.2672253119963 seconds
---------------------------------------------------------------------
With fix
---------------------------------------------------------------------
Throughput (including tokenization) = 9801.250572694065 tokens/second
Memory allocated                    = 19.63 GB
Max memory allocated                = 20.96 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.1736303310026415 seconds
---------------------------------------------------------------------

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-7b-hf/ --max_new_tokens 1000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 32 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
----------------------------------------------------------------------
Throughput (including tokenization) = 1865.7528637898442 tokens/second
Memory allocated                    = 31.03 GB
Max memory allocated                = 31.83 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 49.80568245299219 seconds
----------------------------------------------------------------------
With fix
----------------------------------------------------------------------
Throughput (including tokenization) = 2022.6046518376145 tokens/second
Memory allocated                    = 28.42 GB
Max memory allocated                = 29.18 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 17.86547275799967 seconds
----------------------------------------------------------------------

**Comment by puneeshkhanna** [comment]
Llama 13B perf results without and with fix as below:
python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-13b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 1 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 73.41020172978796 tokens/second
Memory allocated                    = 24.35 GB
Max memory allocated                = 24.36 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 4.884657262999099 seconds
With fix
Throughput (including tokenization) = 73.44032878995496 tokens/second
Memory allocated                    = 24.34 GB
Max memory allocated                = 24.35 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.37153663499339 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-13b-hf/ --max_new_tokens 1000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 1 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 75.7109614095682 tokens/second
Memory allocated                    = 25.2 GB
Max memory allocated                = 25.25 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 18.432767632009927 seconds
With fix
Throughput (including tokenization) = 76.26934607236079 tokens/second
Memory allocated                    = 25.06 GB
Max memory allocated                = 25.08 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 15.193954528993345 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-13b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 64 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 2865.2254885524662 tokens/second
Memory allocated                    = 30.45 GB
Max memory allocated                = 31.22 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 17.60540542700619 seconds
With fix
Throughput (including tokenization) = 3633.9957034341414 tokens/second
Memory allocated                    = 29.76 GB
Max memory allocated                = 30.42 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 4.64509762400121 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 1 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-13b-hf/ --max_new_tokens 1000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 16 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 763.3575211375546 tokens/second
Memory allocated                    = 38.5 GB
Max memory allocated                = 38.92 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 49.66154878999805 seconds
With fix
Throughput (including tokenization) = 812.995454113312 tokens/second
Memory allocated                    = 36.65 GB
Max memory allocated                = 37.03 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 22.37336895200133 seconds

**Comment by puneeshkhanna** [comment]
Llama70B results:

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 2 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 110.62860953193044 tokens/second
Memory allocated                    = 16.96 GB
Max memory allocated                = 17.37 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.4168114269996295 seconds
With fix
Throughput (including tokenization) = 109.80677493423512 tokens/second
Memory allocated                    = 16.95 GB
Max memory allocated                = 16.98 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.253933574000257 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 4000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 2 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 116.59096866609697 tokens/second
Memory allocated                    = 18.87 GB
Max memory allocated                = 39.02 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 70.28810204200272 seconds
With fix
Throughput (including tokenization) = 117.39353441901102 tokens/second
Memory allocated                    = 17.54 GB
Max memory allocated                = 17.58 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 69.67671067500487 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 100 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 40 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 1744.4332593278932 tokens/second
Memory allocated                    = 17.71 GB
Max memory allocated                = 31.47 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.674584093008889 seconds
With fix
Throughput (including tokenization) = 1873.2515938931053 tokens/second
Memory allocated                    = 17.33 GB
Max memory allocated                = 17.95 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.530354335991433 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 4000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 40 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache
Without fix
Throughput (including tokenization) = 1241.8834557707662 tokens/second
Memory allocated                    = 50.08 GB
Max memory allocated                = 88.1 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 137.97351068799617 seconds
With fix
Throughput (including tokenization) = 1409.4239974630066 tokens/second
Memory allocated                    = 23.58 GB
Max memory allocated                = 24.4 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 115.37053568501142 seconds

python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path /software/data/llama_inference/Llama-2-70b-hf/ --max_new_tokens 1000 --bf16 --n_iterations 3 --use_hpu_graphs --use_kv_cache --batch_size 16 --limit_hpu_graphs --warmup 1 --attn_softmax_bf16 --trim_logits --reuse_cache --prompt "<300 tokens>"
Without fix
Throughput (including tokenization) = 778.5468898740977 tokens/second
Memory allocated                    = 19.79 GB
Max memory allocated                = 41.1 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 22.690399918996263 seconds
With fix
Throughput (including tokenization) = 831.5857981807274 tokens/second
Memory allocated                    = 18.26 GB
Max memory allocated                = 34.62 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 21.228477083001053 seconds

File: optimum/habana/transformers/generation/configuration_utils.py
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:35]: Having `None` as the default value enables us to know whether it was set by the user or not in other pieces of the codebase. Is it an issue to keep it that way?

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py:35]: https://github.com/huggingface/transformers/blob/v4.33.2/src/transformers/generation/configuration_utils.py#L39 -see in general transformers config too. I think bool default value should be always False. None makes the check of these attributes quite confusing since False is not equivalent to None.

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py:35]: OK I will push/revert this change in some time.

File: optimum/habana/transformers/models/llama/modeling_llama.py
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:36]: @puneeshkhanna what is the real shape of cur, I think the shape of prev is [N, M, max_T, H]

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: See below change which gets executed only when past key values is None or basically for the 1st token.
cur will be set to the actual seq len for the 1st token [N, M, actual seq len, H]

elif reuse_cache and token_idx is not None:
            # With reuse_cache, KV cache is pre allocated hence for the 1st token we can slice the inputs till token idx for the fwd pass
            input_ids = input_ids[:, :token_idx]
            attention_mask = attention_mask[:, :token_idx]

Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: so, if length of prompts between different BS are not same, there will be recompilation for 1st token?
correct me if I'm wrong

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: That is always the case irrespective of my change and even recreation of KV cache. 
I m pushing a separate PR to have an addition argument of max input tokens before hand to avoid such recompilations.

Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: if with max_input, here we can not use idx, we should use max_input_length right?

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: token_idx is always set to the index of the token that needs to be generated. With max input length, tokenizer will pad/truncate the input seq till that max input tokens. For that, change is only in run generation script only.

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: https://github.com/huggingface/optimum-habana/pull/426 - PR for configurable max input tokens to avoid graph recompiles between batches of various lengths.

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: @ZhaiFeiyue - I think I m understanding the concern since token idx may change between batches. Let me try it out once.

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: @ZhaiFeiyue - we are fine. Preferably we should set max input tokens and since padding is always done on left side, token_idx will be always set to the max input token value only during the 1st token generation for any batch of input sequences and hence no recompilations.

File: optimum/habana/transformers/generation/configuration_utils.py
Comment (regisss) [reply optimum/habana/transformers/generation/configuration_utils.py:35]: I agree with you it would be better that way. My only concern is that there might be places in the codebase where this change will have unexpected consequences. Let's address this in another PR if that sounds good to you?

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py:35]: Yes will push revert in a couple of hrs. Lets address in another PR.

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py]: Pushed revert of changing default values from None to False.

**Review by regisss — APPROVED** [review]
LGTM! Let's wait for @ZhaiFeiyue's approval to merge.

@puneeshkhanna The numbers your gave were obtained on Gaudi2 right?

**Comment by puneeshkhanna** [comment]
> LGTM! Let's wait for @ZhaiFeiyue's approval to merge.
> 
> @puneeshkhanna The numbers your gave were obtained on Gaudi2 right?
Yes.

File: optimum/habana/transformers/models/llama/modeling_llama.py
Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: image

see above picture, token_idx is not equal to max_input_length I think

Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: what I think is here,  prev[:, :, :max_input_length, :].copy_(cur)
if user not set max_input_length, I think it should be max_token_length, since with static shape the length of cur is max_token_length

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: No the green padding is on the left side only for decoder text generation otherwise generation won't be correct.  token_idx initialization below to input_tensor.shape[-1].
Am I missing anything ?

# token_idx is the current index in the generation process, it is incremented each time a new token is generated
model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)

Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: For example see dump of input tokens of BS-4 below with max input tokens as say 50 ; padding is always on the left side.

'input_ids': tensor([[    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     1,  7569,   650,   338,  9796,
           322,   306,   508,  7569,   650,   338,  9796,   322,   306,   508],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     1,   450,   716, 14064,   393,  2355, 19054,
           445,  1629,   450,   716, 14064,   393,  2355, 19054,   445,  1629],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     1,
           512,   278,  2215,  2215,  5418,   515,  1749, 15400, 29891, 29892,
           512,   278,  2215,  2215,  5418,   515,  1749, 15400, 29891, 29892],
        [    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     1,
         24819,   338,   278,   871,   982, 24819,   338,   278,   871,   982,
         24819,   338,   278,   871,   982, 24819,   338,   278,   871,   982]])

Comment (ZhaiFeiyue) [reply optimum/habana/transformers/models/llama/modeling_llama.py:36]: thanks @puneeshkhanna, seems that current implement of `max_input_tokens` is different from what I think.
if with the current implement, you are right.
I will merge

File: optimum/habana/transformers/models/llama/modeling_llama.py
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:94]: @puneeshkhanna  `torch.zeros` lead to more memory consumed when running with FP8 in 1.13.0, can you help to double check?

**Pull request merged**
